lib/services/local/generic/common.go

```diff
@@ -0,0 +1,59 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package generic
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// MarshalFunc is a type signature for a marshaling function.
+type MarshalFunc[T any] func(T, ...services.MarshalOption) ([]byte, error)
+
+// UnmarshalFunc is a type signature for an unmarshalling function.
+type UnmarshalFunc[T any] func([]byte, ...services.MarshalOption) (T, error)
+
+// ServiceCommon is a common generic service interface.
+type ServiceCommon[T any] interface {
```

Comment:
Do we really need this interface? Isn't the generic service implementation enough?

Should we instead be encouraging folks to match against smaller, more focused interfaces only where it's necessary?

Reply:
My intention throughout this PR was to expose identical interface for both old and new style resources. I realised, however, that I didn't quite realised my idea fully. I did this now with bda9073d81758368caecf1a16b1ac52f02b609f1, which unexports `Service` struct in favour of a common `Service` interface (renamed from `ServiceCommon`).

I'm all for evolving the generic service to be easier to user and more performant, but I think these concerns are orthogonal with the intent behind this PR, which was to made the generic service compatible with RFD 153-style resources.

Reply:
The added interface pertains to the PR, but much of the rest I agree - these are mainly wishful comments on my part.

Reply:
I've reworked the entire thing, the interface is gone.